### ISSUE TITLE
Exposing a rust library to C code

### ISSUE THREAD

**Issue by SirVer** [bazelbuild/rules_rust#55]
My use case is to swap out unsafe C with save rust implementations. To keep APIs in tact, I need to provide a C library that bridges to Rust. I was trying to do this:

~~~
package(default_visibility = ["//visibility:public"])

load(
    "@io_bazel_rules_rust//rust:rust.bzl",
    "rust_library",
)

rust_library(
    name = "greeter",
    srcs = ["lib.rs"],
)

cc_binary(
   name = "rust_from_c",
   srcs = ["hello_world.cc"],
   deps = [":greeter"],
)
~~~

with `lib.rs`:
~~~
#[no_mangle]
pub extern "C" fn greet(){
    println!("Hello from rust!");
}
~~~

and `hello_world.cc`:
~~~
void greet();
int main(int argc, char *argv[])
{
	greet();
	return 0;
}
~~~

This should conceptually work, but bazel refuses:
~~~
ERROR: /usr/src/app/rust_from_c/BUILD:18:14: in deps attribute of cc_library rule //rust_from_c:greeter: rust_library rule '//rust_from_c:greeter_rust' is misplaced here (expected cc_inc_library, cc_library, objc_library or cc_proto_library)
~~~

Is there a way to call rust code from C?

**Comment by SirVer** [comment]
@acmcarther I think you know most about these rules, right?

**Comment by acmcarther** [comment]
From my initial research:

https://doc.rust-lang.org/reference/linkage.html
You either want "staticlib" or "cdylib". Specifying crate types is enabled in #38, which was historically blocked on CI issues (I need to fix the PR urgently).

Once that is done, it is possible to pass this file output into the `srcs` of a cc_library or cc_binary rule. See the bazel docs for `cc_library#srcs`: https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library.

This may not work (either because the providers from the rust_library aren't _quite_ right, or because `cc_library#srcs` requires inputs to be statically known (not generated via another rule). In that case, we'd want to consider adding a `cc_library_from_rust_library` rule that reshapes the providers yielded by `rust_library` into a form that `cc_library` will accept.

**Comment by SirVer** [comment]
Thank you! So this is indeed blocked on #38. Also, having to state a `rust_library` as a `src` for a `cc_library` seems odd - from the semantics, `dep` would be the better keyword. But if `src` works I'll take it.

**Comment by mfarrugi** [comment]
Added an example of this here [c_calling_rust](https://github.com/mfarrugi/rules_rust/blob/marco-c-calling-rust-example/examples/c_calling_rust/BUILD), can tidy it up and create a pr after #38 

Can't currently invoke the cc_library rule from rust_library like a macro can, so we could add a macro that calls `rust_library` and then `cc_library as in the example or figure out what it would take for cc_* rules to be more flexible in accepting native dependencies.

**Comment by SirVer** [comment]
awesome! 

/cc @thomasschiwietz

**Comment by mfarrugi** [comment]
This is now possible and an example is pushed.
Dylibs are a little trickier but support for them on linux is pending.

**Comment by thomasschiwietz** [comment]
very cool!

**Comment by mfarrugi** [comment]
Resolved by #74
